### ISSUE TITLE
error MT4109: Failed to compile the generated registrar code.

### ISSUE THREAD

**Issue by rvpeepsdev** [xamarin/xamarin-macios#16007]
### Steps to Reproduce

1. Create Xamarin Forms app using any of the three templates
2. Connect to remote Mac 
3. Set platform to iOS/iPhone and pick physical phone connected to MAC
4. Build

### Expected Behavior
Successful build

### Actual Behavior
build failed with

```
C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(235,3): error MT4109: Failed to compile the generated registrar code. Please file a bug report at https://github.com/xamarin/xamarin-macios/issues/new
========== Rebuild All: 1 succeeded, 1 failed, 0 skipped ==========
```

### Environment
Windows 11 Pro 21H2:
- VisualStudio 2022 v17.3.4

MacOS Monterey 12.6: 
- Xcode 14.0 (14A309)
- VisualStudio 17.4 (build 1552) (also tried build 715)

### Build Logs
```
1>------ Rebuild All started: Project: DXCollectionViewTest, Configuration: Debug Any CPU ------
Restored D:\scratch\DXCollectionViewTest\DXCollectionViewTest\DXCollectionViewTest.csproj (in 6 ms).
Restored D:\scratch\DXCollectionViewTest\DXCollectionViewTest.iOS\DXCollectionViewTest.iOS.csproj (in 8 ms).
Restored D:\scratch\DXCollectionViewTest\DXCollectionViewTest.Android\DXCollectionViewTest.Android.csproj (in 22 ms).
1>DXCollectionViewTest -> D:\scratch\DXCollectionViewTest\DXCollectionViewTest\bin\Debug\netstandard2.0\DXCollectionViewTest.dll
2>------ Rebuild All started: Project: DXCollectionViewTest.iOS, Configuration: Debug iPhone ------
2>  Executing SayHello Task to establish a connection to a Remote Server. 
2>  			Properties: 
2>  				SessionId=xxx, 
2>  				Addresss=xxx, 
2>  				SshPort=22, 
2>  				TcpPort=52822, 
2>  				User=xxx, 
2>  				AppName=DXCollectionViewTest.iOS,
2>  				VisualStudioProcessId=26396,
2>  				ContinueOnDisconnected=False
2>  Detected signing identity:
2>          
2>    Code Signing Key: "Apple Development: Created via API (xxx)" (xxx)
2>    Provisioning Profile: "VS: WildCard Development" (xxx)
2>    Bundle Id: com.companyname.DXCollectionViewTest
2>    App Id: xxx.com.companyname.DXCollectionViewTest
2>  DXCollectionViewTest.iOS -> D:\scratch\DXCollectionViewTest\DXCollectionViewTest.iOS\bin\iPhone\Debug\DXCollectionViewTest.iOS.exe
2>  /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/bin/mtouch @/Users/xxx/Library/Caches/Xamarin/mtbs/builds/DXCollectionViewTest.iOS/xxx/obj/iPhone/Debug/response-file.rsp 
2>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(235,3): error MT4109: Failed to compile the generated registrar code. Please file a bug report at https://github.com/xamarin/xamarin-macios/issues/new
========== Rebuild All: 1 succeeded, 1 failed, 0 skipped ==========
```

**Comment by rvpeepsdev** [comment]
fwiw, using @dalexsoto suggestion of setting the linker to [Link SDK assemblies only](https://docs.microsoft.com/en-us/xamarin/ios/deploy-test/linker?tabs=macos#link-sdk-assemblies-only), did get rid of the registrar errors in this instance.

link to similar discussion thread is https://github.com/xamarin/xamarin-macios/issues/15954

**Comment by lucapan** [comment]
Same problem here! :(

**Comment by tj-devel709** [comment]
Hello @rvpeepsdev. Thank you for your feedback. Unfortunately setting the linker to Link SDK assemblies only works as a temporary solution for some instances but not all. In your case, I would continue to monitor the xcode 14 issue found [here](https://github.com/xamarin/xamarin-macios/issues/15954) and wait for that update or revert to xcode13.4.1 for the time being. Thanks

**Comment by clagoglez** [comment]
Hello guys, just to let you know that I have (temporary) solved this issue by uninstalling xCode 14.0 and installing xCode 13.4.1 again, like @tj-devel709 mentioned above. The only thing is that after the new xCode installation, you need to reload your Apple Developer Account info, and in VS for Mac, you need to enter the AppleStoreConnect API Key. If you don't know on how to proceed, follow the link on this screen in VS for Mac. I hope this help you out ;-)
![image](https://user-images.githubusercontent.com/13588831/191796229-0d8ab09b-446a-4791-b0c9-c7ca73dd6264.png)

**Comment by RobTF** [comment]
Looks like this is back and we're currently blocked on a production release :/. For us it seems to occur on release build only with the presence of an iOS extension project.

**Comment by RobTF** [comment]
In case it helps this is what I'm getting;

```
		
5>  		
5>  Compilation failed with code 1, command:
5>  /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -stdlib=libc++ -gdwarf-2 -std=c++14 -I/Library/Frameworks/Xamarin.iOS.framework/Versions/16.0.0.75/SDKs/MonoTouch.iphoneos.sdk/include -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.2.sdk -Qunused-arguments -miphoneos-version-min=11.0 -arch arm64 -c -o /Users/rob/Library/Caches/Xamarin/mtbs/builds/My.App.Extension.MyWidget/76367a930e3be44d8dd3902c846e945ec5a73078f926161804f5d1e4dab2de8d/obj/iPhone/Release/mtouch-cache/arm64/main.o -x objective-c++ -D MONOTOUCH /Users/rob/Library/Caches/Xamarin/mtbs/builds/My.App.Extension.MyWidget/76367a930e3be44d8dd3902c846e945ec5a73078f926161804f5d1e4dab2de8d/obj/iPhone/Release/mtouch-cache/arm64/main.m
5>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(235,3): warning : no such sysroot directory: '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.2.sdk' [-Wmissing-sysroot]
5>  In file included from /Users/rob/Library/Caches/Xamarin/mtbs/builds/My.App.Extension.MyWidget/76367a930e3be44d8dd3902c846e945ec5a73078f926161804f5d1e4dab2de8d/obj/iPhone/Release/mtouch-cache/arm64/main.m:8:
5>  In file included from /Library/Frameworks/Xamarin.iOS.framework/Versions/16.0.0.75/SDKs/MonoTouch.iphoneos.sdk/include/xamarin/xamarin.h:12:
5>  /Library/Frameworks/Xamarin.iOS.framework/Versions/16.0.0.75/SDKs/MonoTouch.iphoneos.sdk/include/xamarin/main.h:13:10: fatal error: 'Foundation/Foundation.h' file not found
5>  #include <Foundation/Foundation.h>
5>           ^~~~~~~~~~~~~~~~~~~~~~~~~
5>  1 error generated.
5>  Compilation failed with code 1, command:
5>  /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -Wno-receiver-forward-class -Wno-objc-missing-super-calls -Wno-unguarded-availability-new -stdlib=libc++ -gdwarf-2 -std=c++14 -I/Library/Frameworks/Xamarin.iOS.framework/Versions/16.0.0.75/SDKs/MonoTouch.iphoneos.sdk/include -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.2.sdk -Qunused-arguments -miphoneos-version-min=11.0 -arch arm64 -c -o /Users/rob/Library/Caches/Xamarin/mtbs/builds/My.App.Extension.MyWidget/76367a930e3be44d8dd3902c846e945ec5a73078f926161804f5d1e4dab2de8d/obj/iPhone/Release/mtouch-cache/arm64/registrar.o -x objective-c++ /Users/rob/Library/Caches/Xamarin/mtbs/builds/My.App.Extension.MyWidget/76367a930e3be44d8dd3902c846e945ec5a73078f926161804f5d1e4dab2de8d/obj/iPhone/Release/mtouch-cache/registrar.m
5>C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Xamarin\iOS\Xamarin.iOS.Common.targets(235,3): warning : no such sysroot directory: '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.2.sdk' [-Wmissing-sysroot]
5>  In file included from /Users/rob/Library/Caches/Xamarin/mtbs/builds/My.App.Extension.MyWidget/76367a930e3be44d8dd3902c846e945ec5a73078f926161804f5d1e4dab2de8d/obj/iPhone/Release/mtouch-cache/registrar.m:1:
5>  In file included from /Library/Frameworks/Xamarin.iOS.framework/Versions/16.0.0.75/SDKs/MonoTouch.iphoneos.sdk/include/xamarin/xamarin.h:12:
5>  /Library/Frameworks/Xamarin.iOS.framework/Versions/16.0.0.75/SDKs/MonoTouch.iphoneos.sdk/include/xamarin/main.h:13:10: fatal error: 'Foundation/Foundation.h' file not found
5>  #include <Foundation/Foundation.h>
5>           ^~~~~~~~~~~~~~~~~~~~~~~~~
5>  1 error generated.
```

**Comment by RobTF** [comment]
I just noticed - the build process is looking for this path
`/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.2.sdk`

15.2 SDK?? I'm set to 16.0? Interestingly due to pathing issues I have a symlink to my project folder which I use occasionally for other projects in the repo. I tried using the symlink and the version of this SDK folder changed to 15.4. Is there some sort of cache somewhere that's affecting stuff? I've looked throughout the project and can't find any traces of iOS 15 SDK - everything is version 16.
